README.md
# tp-logger

Supplement log messages with a Timestamp and caller's Path.

## Install

`$ npm install -save tp-logger`

## Usage

```javascript
var logger = require('tp-logger')(/* optional options, see Options below */);

// Behaves the same as default console, where you can use string formatting
logger.log('lol: %d hours', 2);
// [2015-12-16T06:49:08+00:00][vanilla.js] lol: 2 hours

logger.info('dota');
// [2015-12-16T06:49:08+00:00][vanilla.js] dota

logger.error(new Error('helldivers').stack);
// [2015-12-16T06:49:08+00:00][vanilla.js] Error: helldivers
//     at Object.<anonymous> (/Users/woozyking/proj/personal/tp-logger/examples/vanilla.js:9:14)
//     at Module._compile (module.js:425:26)
//     at Object.Module._extensions..js (module.js:432:10)
//     at Module.load (module.js:356:32)
//     at Function.Module._load (module.js:313:12)
//     at Function.Module.runMain (module.js:457:10)
//     at startup (node.js:138:18)
//     at node.js:974:3
```

If you don't mind your coworkers wondering why node.js `console` has suddenly become so awesome (or bad), you can also do:

```javascript
// NOT Recommended
console = require('tp-logger')(/* optional options, see Options below */);
// ...
```

## Options

See them in action by running [this example](examples/options.js)

```javascript
var logger = require('tp-logger')({
  // stream for log() and info()
  stdout: <writable stream instances. Default: process.stdout>,

  // stream for error()
  stderr: <writable stream instances. Default: process.stderr>,

  // whether to use local time or UTC time
  local: <boolean. Default: false>,
  localTs: <Alias of 'local'. Since 0.2>,

  // timestamp format
  tsFormat: <string. See moment.js formatting documentation>,

  // whether to log full path of the caller or just caller filename
  fullPath: <boolean. Default: false. Since 0.2>,

  // whether to log with a third supplement of [log|info|err]
  // default to true when stdout == stderr, otherwise false
  logType: <boolean. Default: depends on whether stdout == stderr. Since 0.2>
});
```


CHANGELOG
# v0.2

* __Adds__ `fullPath` option. When `true`, messages logged are supplemented by caller's full path rather than just "base name" (module's file name). Defaults to `false`.
* __Adds__ `localTs` option, alias of `local` option for now, and in future versions will be the preferred one as `local` is ambiguous.
* __Adds__ new behavior when both `stdout` and `stderr` point to the same stream, in which case messages logged contain a third supplement of [`info|log|error`] log type specification.
* __Adds__ `logType` option. When `true`, logger behaves the same as when both `stdout` and `stderr` point to the same stream. Default depends on whether `stdout` and `stderr` point to the same stream or not.


index.js
// node.js built-in
var path = require('path');
var format = require('util').format;
var Console = require('console').Console;
// 3rd party
var moment = require('moment');

/*
  Module entry point responsible for taking in some options to adjust behavior

  Example
  opt = {
    stdout: <writable stream instances. Default: process.stdout>,
    stderr: <writable stream instances. Default: process.stderr>,
    local: <boolean. Default: false>,
    localTs: <Alias of 'local'. Since 0.2>,
    tsFormat: <string. See moment.js formatting documentation>,
    fullPath: <boolean. Default: false. Since 0.2>,
    logType: <boolean. Default: depends on whether stdout == stderr. Since 0.2>
  }
*/
var tpLogger = (function(opt) {
  opt = opt || {};
  opt.stdout = opt.stdout || process.stdout;
  opt.stderr = opt.stderr || process.stderr;

  // This is nothing but a simple duplicate of standard console found in node.js
  var logger = new Console(opt.stdout, opt.stderr);

  // override the following console methods
  ['log', 'info', 'error'].forEach(function(method) {
    // original method
    var orig = logger[method].bind(logger);

    // add a healthy dose of supplements to log messages
    logger[method] = function() {
      // supplement of time
      var ts = ((opt.local || opt.localTs) ? moment() : moment.utc()).format(opt.tsFormat);
      // supplement of path
      var callerPath = opt.fullPath ? process.mainModule.filename : path.basename(process.mainModule.filename);
      // assemble all supplements
      var supp = format('[%s][%s]', ts, callerPath);

      // if logType option is true, or when stdout and stderr are the same
      if (opt.logType || (opt.stdout == opt.stderr)) {
        // supplement of log type
        supp += format('[%s]', method);
      }

      // Prepend supplements, followed by original log message
      orig.apply(logger, [
        supp,
        format.apply(format, Array.prototype.slice.call(arguments))
      ]);
    };
  });

  return logger;
});

module.exports = tpLogger;


examples/options.js
/*
  examples/options.js demonstrate options' impact on log messages.
*/
var fs = require('fs');

var stdout = fs.createWriteStream('./stdout.log');
var stderr = fs.createWriteStream('./stderr.log');
var logger = require('../index.js')({
  stdout: stdout,
  stderr: stderr,
  localTs: true,
  tsFormat: 'dddd, MMMM Do YYYY, h:mm:ss a Z',
  fullPath: true,
  logType: true
});

logger.log('lol: %d hours', 2); // in ./stdout.log
// [Friday, December 18th 2015, 1:59:28 am -05:00][/Users/woozyking/proj/personal/tp-logger/examples/options.js][log] lol: 2 hours

logger.info('dota'); // in ./stdout.log
// [Friday, December 18th 2015, 1:59:28 am -05:00][/Users/woozyking/proj/personal/tp-logger/examples/options.js][info] dota

logger.error(new Error('helldivers').stack); // in ./stderr.log
// [Friday, December 18th 2015, 1:59:28 am -05:00][/Users/woozyking/proj/personal/tp-logger/examples/options.js][error] Error: helldivers
//     at Object.<anonymous> (/Users/woozyking/proj/personal/tp-logger/examples/options.js:23:14)
//     at Module._compile (module.js:398:26)
//     at Object.Module._extensions..js (module.js:405:10)
//     at Module.load (module.js:344:32)
//     at Function.Module._load (module.js:301:12)
//     at Function.Module.runMain (module.js:430:10)
//     at startup (node.js:141:18)
//     at node.js:980:3


examples/vanilla.js
/*
  examples/vanilla.js demonstrate out of the box vanilla experience you get from
  tp-logger
*/
var logger = require('../index.js')();

logger.log('lol: %d hours', 2);
// [2015-12-16T06:49:08+00:00][vanilla.js] lol: 2 hours

logger.info('dota');
// [2015-12-16T06:49:08+00:00][vanilla.js] dota

logger.error(new Error('helldivers').stack);
// [2015-12-16T06:49:08+00:00][vanilla.js] Error: helldivers
//     at Object.<anonymous> (/Users/woozyking/proj/personal/tp-logger/examples/vanilla.js:9:14)
//     at Module._compile (module.js:425:26)
//     at Object.Module._extensions..js (module.js:432:10)
//     at Module.load (module.js:356:32)
//     at Function.Module._load (module.js:313:12)
//     at Function.Module.runMain (module.js:457:10)
//     at startup (node.js:138:18)
//     at node.js:974:3
